## Lambda, Filter, Map, Reduce

---

## Lambda Functions

Давайте совместно переконвертируем функцию из Python в lambda выражение.

Для понимания синтаксиса lambda, мы будем использовать следущий пример:

![lambda_diagram.png](lamda_diagram.png)

### Q 1: Простая функция

Переделаем обычную функцию в lambda

In [4]:
def add_nums(x, y):
    return x + y

In [ ]:
# напишите lambda выражение, которая будет работать одинаково с add_nums

add_nums_lmda = 

In [6]:
# Тест на равность результатов
assert add_nums_lmda(5,10) == add_nums(5,10), "Your function is incorrect"
assert add_nums_lmda(-5,10) == add_nums(-5,10), "Your function is incorrect"
assert add_nums_lmda(100,-10) == add_nums(-10,100), "Your function is incorrect"

### Q 2: Max Function

Переделаем фунцию, которая сравнивает два числа и выводит наибольшее. Пример if условия в lambda:

![if_else_expression.png](if_else_expression.png)

In [ ]:
def max_num(x,y):
    if x > y:
        return x
    else:
        return y

In [ ]:
# напишите lambda выражение, которая будет работать одинаково с max_num

max_num_lmda = 

In [ ]:
# Тест
assert max_num_lmda(50,10) == max_num(50,10), "Your function is incorrect"
assert max_num_lmda(10,50) == max_num(10,50), "Your function is incorrect"
assert max_num_lmda(50,50) == max_num(50,50), "Your function is incorrect"

---
## List Comprehension

Напомним list comprehensions (рассмотрим на примере)

### Q 3: Even Numbers

In [9]:
# создайте список из чисел от 2 до 200, которые делятся на 2 без остатка

even_numbers = 

print(even_numbers)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190, 192, 194, 196, 198, 200]


In [10]:
# Testing your list
def test_even_list(lst):
    return list(range(2,201,2)) == lst

test_even_list(even_numbers)

True

### Q 4: Powers of 2

In [ ]:
# создайте список из чисел от 2**0 до 2**50

powers_of_two = 

print(powers_of_two)

In [ ]:
# Testing your list
def test_powers_of_two_list(lst):
    return [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 
            2048, 4096, 8192, 16384, 32768, 65536, 131072, 
            262144, 524288, 1048576, 2097152, 4194304, 8388608, 
            16777216, 33554432, 67108864, 134217728, 268435456, 
            536870912, 1073741824, 2147483648, 4294967296, 8589934592, 
            17179869184, 34359738368, 68719476736, 137438953472, 274877906944, 
            549755813888, 1099511627776, 2199023255552, 4398046511104, 
            8796093022208, 17592186044416, 35184372088832, 70368744177664, 
            140737488355328, 281474976710656, 562949953421312, 1125899906842624] == lst

test_powers_of_two_list(powers_of_two)

---
## Filter function

Научимся создавать фильтры

### Q 5: Only Vowels

Сделайте функцию, которая сделает фильтр и вернет только гласные буквы в нижнем регистре.

Пример input `['t', 'P', 'R', 'Y', 'L', 'U', 'B', 'l', 'N', 'J', 's', 'C', 'p', 'o', 'M', 
            'x', 'u', 'n', 'q', 'B', 'h', 'W', 'C', 'S', 'L', 'X', 'f', 'T', 'x', 'R', 
            'a', 'j', 'F', 'J', 'h', 'L', 'I', 'J', 'L', 'W', 'Q', 'S', 'G', 'B', 'P', 
            'q', 'S', 'Q', 'Q', 'o', 'V', 'L', 'V', 'O', 'G', 'K', 'm', 'Y', 'c', 'k', 
            'v', 'b', 't', 'h', 'C', 'B', 'k', 'p', 'J', 'N', 'o', 'l', 'R', 'a', 'e', 
            'S', 'J', 'Q', 'n', 'b', 'g', 'x', 'g', 'K', 'Y', 't', 'e', 'F', 'i', 'T', 
            'h', 'X', 'o', 'S', 'N', 'l', 'Z', 'U', 'G', 'y']`

Пример output `['o', 'u', 'a', 'o', 'o', 'a', 'e', 'e', 'i', 'o']`

In [ ]:
sequence = ['t', 'P', 'R', 'Y', 'L', 'U', 'B', 'l', 'N', 'J', 's', 'C', 'p', 'o', 'M', 
            'x', 'u', 'n', 'q', 'B', 'h', 'W', 'C', 'S', 'L', 'X', 'f', 'T', 'x', 'R', 
            'a', 'j', 'F', 'J', 'h', 'L', 'I', 'J', 'L', 'W', 'Q', 'S', 'G', 'B', 'P', 
            'q', 'S', 'Q', 'Q', 'o', 'V', 'L', 'V', 'O', 'G', 'K', 'm', 'Y', 'c', 'k', 
            'v', 'b', 't', 'h', 'C', 'B', 'k', 'p', 'J', 'N', 'o', 'l', 'R', 'a', 'e', 
            'S', 'J', 'Q', 'n', 'b', 'g', 'x', 'g', 'K', 'Y', 't', 'e', 'F', 'i', 'T', 
            'h', 'X', 'o', 'S', 'N', 'l', 'Z', 'U', 'G', 'y']



def only_vowels(letter):
    # используя in сделайте сильтр
    # исключите "y"
    pass

# примените функцию в lambda выражении
filter_example = 

print(filter_example)

---
## Map function

Теперь обратимся к функции Map


![map_diagram.png](map_diagram.png)

### Q 6: Squares

Используем map для возведения в квадрат последовательности `sequence_lst = [1,2,7,4,0,6,14,8,9,10]`, результат должен быть `[1, 4, 49, 16, 0, 36, 196, 64, 81, 100]`.

In [ ]:
sequence_lst = [1,2,7,4,0,6,14,8,9,10]

# Add your map function here (hint: you may want to use a lambda!)
map_example = 

print(map_example)

#### Bonus: А можно сделать `map_example`, но с помощью списка?

In [ ]:
# решение
map_example_bonus = 

print(map_example_bonus)

### Q 7: Compute the Average

Реализуем map функцию, которая будет применима для каждого элемента из списка списков `to_average_seq` 

Нам небходимо написать функцию, которая посчитает средние значения по всем вложенным спискам в `to_average_seq`. 

Результат должен быть `[13.0, 47.2, 1.0, 35.8, 37.75]`. 


In [ ]:
to_average_seq = [[0,3,11,45,6],[90,31,44,68,3],[1,1,1,1],[2,5,8,111,53],[99,22,11,19]]

# решение
average_map_example = 

print(average_map_example)

---
## Reduce Function


![reduce_diagram.png](reduce_diagram.png)

### Q 8: Multiply all Elements

Используя функцию reduce сделайте перемножение элементов из последовательность `test_sequence`.

Ваш результат должен быть `360` и функция должна работать, как последовательность `(((3*4)*5)*6) = 360`

Hint: Используем lambda функцию

In [ ]:
test_sequence = [3,4,5,6]

# результат
multiply_all_elements_example = 

print(multiply_all_elements_example)

### Q 9: List of Digits to Number

Преобразуйте список `[1,2,3,4]` в `int`, который будет создержать значение `1234`. 

`[1,2,3,4] -> 1234` где `->` reduce функция

Hint: Подумайте, как использовать `str()` и `int()`. 

In [ ]:
# список
digits_lst = [1,2,3,4]

# результат
digit_lst_to_number = 

print(digit_lst_to_number)
